dispatch.go

```diff
@@ -124,30 +124,33 @@ func (d *dispatch) handleCall(calls map[seqNumber]*call, c *call) {
 	seqid := d.nextSeqid()
 	c.seqid = seqid
 	calls[c.seqid] = c
-	err := d.dispatchMessage(c.ctx, MethodCall, seqid, c.method, c.arg)
-	if err != nil {
-		setResult := c.Finish(err)
-		if !setResult {
-			panic("c.Finish unexpectedly returned false")
-		}
-		return
-	}
+	errCh := d.dispatchMessage(c.ctx, MethodCall, seqid, c.method, c.arg)
 	d.log.ClientCall(seqid, c.method, c.arg)
 	go func() {
-		select {
-		case <-c.ctx.Done():
-			setResult := c.Finish(newCanceledError(c.method, seqid))
-			if !setResult {
-				// The result of the call has already
-				// been processed.
-				return
+		for {
```

Comment:
Okay, now I know why I was confused by this. This control flow is different from the one in my comment https://github.com/keybase/go-framed-msgpack-rpc/issues/26#issuecomment-157615799 -- there's a goroutine here. But I think we don't need that goroutine to do what we want. To expand on my comment in the link above, the control flow should look like:

```
func ... handleCall(...) {
  ...
  errCh := d.dispatchMessage(...)
  ...
  select {
    case <-c.ctx.Done():
      cancelErrCh := ...
      // TODO: I wonder if we shouldn't use c.ctx.Err() here, perhaps as a parameter to newCanceledError.
      setResult := c.Finish(...)
      if !setResult { panic }
      // remove c from calls directly
      // spawn goroutine to wait on cancelErrCh and log it
    case err := <-errCh:
      if err != nil {
        setResult := c.Finish(err)
        if !setResult { panic }
        return
      }
      // Or just add c to calls here instead of by the caller.
  }
}
```

This way, we don't have to select on c.doneCh, since the only other place Finish gets called happens only after a future iteration of callLoop, and for the same reason, setResult should always be true. Also, since we're not in a goroutine, we can then just remove c from calls directly. (In fact, you can just not add it until the nil err from errCh case.)

So in general, we don't need handleCall, Notify, etc. to be completely asynchronous -- we just need it to exit immediately if ctx is cancelled. Which the above does in the simplest way, I think.